pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,244 @@
+	if err != nil {
+		return allowErr(logger, "can't decode old object, allowing the change", err)
+	}
+	logger = log.With(logger, "request_gvk", oldGVK)
+
+	oldReplicas, err := replicas(oldObj, oldGVK)
+	if err != nil {
+		return allowErr(logger, "can't get old replicas, allowing the change", err)
+	}
+	logger = log.With(logger, "old_replicas", int32PtrStr(oldReplicas))
+
+	newObj, newGVK, err := codecs.UniversalDeserializer().Decode(ar.Request.Object.Raw, nil, nil)
+	if err != nil {
+		return allowErr(logger, "can't decode new object, allowing the change", err)
+	}
+
+	newReplicas, err := replicas(newObj, newGVK)
+	if err != nil {
+		return allowErr(logger, "can't get new replicas, allowing the change", err)
+	}
+	logger = log.With(logger, "new_replicas", int32PtrStr(newReplicas))
+
+	// Both replicas are nil, nothing to warn about.
+	if oldReplicas == nil && newReplicas == nil {
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	// Changes from/to nil scale are not downscales strictly speaking.
+	if oldReplicas == nil || newReplicas == nil {
+		return allowWarn(logger, "old/new replicas is nil, allowing the change")
+	}
+	// If it's not a downscale, just log debug.
+	if *oldReplicas < *newReplicas {
+		level.Debug(logger).Log("msg", "upscale allowed")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	if *oldReplicas == *newReplicas {
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	// Get the resource labels: for example, for a StatefulSet, it will be the labels of the StatefulSet itself,
+	// while for a Scale object, it will be the labels of the StatefulSet/Deployment/ReplicaSet that the Scale object belongs to.
+	var lbls map[string]string
+	switch o := oldObj.(type) {
+	case *appsv1.Deployment:
+		lbls = o.Labels
+	case *appsv1.StatefulSet:
+		lbls = o.Labels
+	case *appsv1.ReplicaSet:
+		lbls = o.Labels
+	case *autoscalingv1.Scale:
+		lbls, err = getResourceLabels(ctx, ar, api)
+		if err != nil {
+			return allowBecauseCannotGetResource(ar, logger, err)
+		}
+	default:
+		return allowWarn(logger, fmt.Sprintf("unsupported type %T, allowing the change", o))
+	}
+
+	if lbls[PrepDownscaleLabelKey] != PrepDownscaleLabelValue {
+		// Not labeled, nothing to do.
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	port := lbls[PrepDownscalePortKey]
+	if port == "" {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePortKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePortKey))
+		return &reviewResponse
```

Comment:
Nit: during this last review, I found it hard to follow the code with all the noise of these denied responses. I would consider not introducing a var for them, and defining a method that would format the `Allowed: false` response, like this:

```suggestion
		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePortKey))
		return deny("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePortKey)

// ...

// deny returns a *v1.AdmissionResponse with Allowed: false and the message provided formatted with as in fmt.Sprintf.
func deny(msg, args ...any) *v1.AdmissionResponse {
	return v1.AdmissionResponse{
		Allowed: false,
		Result: &metav1.Status{
			Message: fmt.Sprintf(msg, args...),
		},
	}
}
```

IMO this would make this entire method shorter and easier to follow.

Reply:
Good idea. The function was added manually.